library/src/main/java/bio/terra/pfb/PfbReader.java

```diff
@@ -0,0 +1,182 @@
+import java.net.URLConnection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.specific.SpecificDatumReader;
+
+public class PfbReader {
+
+  public String showSchema(String fileLocation) throws IOException {
+    Schema schema = getSchema(fileLocation);
+    List<Schema> pfbSchemaShow =
+        schema.getField("object").schema().getTypes().stream()
+            .filter(t -> !t.getName().equals("Metadata"))
+            .toList();
+    String show = pfbSchemaShow.stream().map(s -> s.toString()).toList().toString();
+    return convertEnum(show);
+  }
+
+  public String showNodes(String fileLocation) throws IOException {
+    Metadata metadata = deserializeAvroFile(fileLocation);
+    return metadata.getNodes().stream().map(n -> n.getName()).collect(Collectors.joining("\n"))
+        + "\n";
+  }
+
+  public String show(String fileLocation) throws IOException {
+    // TODO - figure out how to read the rest of the Entity objects into a list of single json
+    // objects
+    return "NOT IMPLEMENTED";
+  }
+
+  public String showMetadata(String fileLocation) throws IOException {
+    Metadata metadata = deserializeAvroFile(fileLocation);
+    return metadata.toString();
+  }
+
+  public Schema getSchema(String fileLocation) throws IOException {
+    boolean isUrl = isValidUrl(fileLocation);
+    if (isUrl) {
+      return readUrlPFBSchema(fileLocation);
+    }
+    return readFilePathPFBSchema(fileLocation);
+  }
+
+  Schema readFilePathPFBSchema(String fileLocation) throws IOException {
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    DataFileReader<Entity> dataFileReader =
+        new DataFileReader<>(new File(fileLocation), datumReader);
+    return dataFileReader.getSchema();
+  }
+
+  public Metadata deserializeAvroFile(String fileLocation) throws IOException {
+    File pfbData = new File(fileLocation);
+    // Deserialize the above generated avro data file
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    DataFileReader<Entity> dataFileReader = new DataFileReader<>(pfbData, datumReader);
+    Entity data = null;
+    Metadata result = null;
+    // A PFB Avro file consists of a list of "Entity" objects (Defined in sample.advl)
+    // One of these entities must be a "Metadata" object (Also defined in sample.advl)
+    // The rest of the entities are the data entries
+    // TODO: Is it possible to get in an infinite loop here if there is no metadata object?
+    while (dataFileReader.hasNext()) {
+      // One of the elements must be a "Metadata" object. From this, we can get the schema for the
+      // data
+      try {
+        data = dataFileReader.next(data);
+        result = (Metadata) data.getObject();
+        if (result != null) {
```

Comment:
Just expanding on the TODO here - I ran through the test and it looked like every one of the test files got a result from the first item from the dataFileReader iteration - so could be good to have test files that don't have a metadata object at all as well as not the first item.

Reply:
After reading more about this, I think we can just throw an exception if the first entry is not a metadata entity. [The spec requires the first entity meets the specs for a Metadata entity](https://github.com/uc-cdis/pypfb/blob/master/docs/index.md#schema).